### PR TITLE
Implement 1 to master

### LINKED ISSUE #1: There should be a GitHub OAuth component
**Is your feature request related to a problem? Please describe.**
Currently there is now component for handling GitHub OAuth.

**Describe the solution you'd like**
A component should be created, that can handle GitHub OAuth requests.

### PR DESCRIPTION
**Pull request by pinterid**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Documentation (non-breaking change which adds documentation)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently there is now GitHub OAuth component. #1

**What is the new behavior (if this is a feature change)?**
- Now there is a GitHub OAuth component. Closes #1

**Other information**:
- Ref: https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,33 @@
 {
   "name": "reactjs-oauth",
   "version": "0.1.0",
-  "private": true,
+  "description": "A neat OAuth service for React",
```
Comment (schettn) [inline package.json]:
```suggestion
  "description": "A neat OAuth Service for ReactJS",
```

File: src/GitHub/index.jsx [hunk 2]
```diff
@@ -0,0 +1,158 @@
+//#region > Imports
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+//> MDB
+// "Material Design for Bootstrap" is a great UI design framework
+import { MDBBtn, MDBIcon } from "mdbreact";
+
+//> PopupWindow
+import PopupWindow from "../PopupWindow";
```
Comment (schettn) [inline src/GitHub/index.jsx:13]: What is a popup window?

File: src/GitHub/index.jsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ * @class A button that is used for GitHub OAuth
+ */
+class OAuth2Login extends Component {
+  constructor(props) {
+    super(props);
+    this.onBtnClick = this.onBtnClick.bind(this);
+    this.onRequest = this.onRequest.bind(this);
+    this.onSuccess = this.onSuccess.bind(this);
+    this.onFailure = this.onFailure.bind(this);
+  }
```
Comment (schettn) [inline src/GitHub/index.jsx]:
```suggestion
  constructor(props) {
    super(props);
    
    this.onBtnClick = this.onBtnClick.bind(this);
    this.onRequest = this.onRequest.bind(this);
    this.onSuccess = this.onSuccess.bind(this);
    this.onFailure = this.onFailure.bind(this);
  }
```

File: src/GitHub/index.jsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      clientSecret,
+      scope,
+      redirectUri,
+    } = this.props;
+    const search = toQuery({
+      client_id: clientId,
+      scope,
+      redirect_uri: redirectUri,
+      response_type: "code",
+    });
```
Comment (schettn) [inline src/GitHub/index.jsx]:
```suggestion
    const {
      authorizationUrl,
      clientId,
      clientSecret,
      scope,
      redirectUri,
    } = this.props;
    
    const search = toQuery({
      client_id: clientId,
      scope,
      redirect_uri: redirectUri,
      response_type: "code",
    });
```

File: src/GitHub/index.jsx [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    const popup = PopupWindow.open(`${authorizationUrl}?${search}`);
+    this.popup = popup;
+
+    this.onRequest();
+    popup.then(
+      (data) => this.onSuccess(data),
+      (error) => this.onFailure(error)
+    );
+  }
```
Comment (schettn) [inline src/GitHub/index.jsx]: Wouldn't it be possible to assign this directly?

`this.popup = PopupWindow.open(`${authorizationUrl}?${search}`);`

Comment (schettn) [reply src/GitHub/index.jsx]: If not, there are some newline issues in this region.

File: src/GitHub/index.jsx [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      return this.onFailure(new Error("'code' not found"));
+    }
+
+    const AuthorizeUrl =
+      `https://cors.snek.at/` +
+      `https://github.com/login/oauth/access_token?code=${data.code}` +
+      `&client_id=${this.props.clientId}` +
+      `&client_secret=${this.props.clientSecret}` +
+      `&redirect_uri=${this.props.redirectUri}` +
+      `&state=${await GuidGenerator()}`;
```
Comment (schettn) [inline src/GitHub/index.jsx]: It seems like the scope is missing. Therefore the requests access token has zero privileges.

image

File: src/GitHub/index.jsx [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+            return onSuccess(data);
+          });
+      });
+  }
+
+  onFailure(error) {
+    const { onRequest } = this.props;
+    console.log(error);
+    onRequest(error);
```
Comment (schettn) [inline src/GitHub/index.jsx]: I would just throw an error. Somebody will catch it for sure.

File: src/GitHub/index.jsx [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    );
+  }
+}
+//#endregion
+
+//#region > Properties
+OAuth2Login.defaultProps = {
+  scope: "",
+  onRequest: () => {},
+};
```
Comment (schettn) [inline src/GitHub/index.jsx]: I would set the default scope to: `user`

You can see all possible scopes on: [GitHub OAuth Scopes](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/)

File: src/PopupWindow.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+          if (
+            popup.location.href === this.url ||
+            popup.location.pathname === "blank"
+          ) {
+            return;
+          }
+
+          const params = toParams(popup.location.search);
+          resolve(params);
```
Comment (schettn) [inline src/PopupWindow.js]:
```suggestion
          const params = toParams(popup.location.search);
          
          resolve(params);
```

File: src/PopupWindow.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  poll() {
+    this.promise = new Promise((resolve, reject) => {
+      this.iid = window.setInterval(() => {
+        try {
+          const popup = this.window;
+
+          if (!popup || popup.closed !== false) {
+            this.close();
+
+            reject(new Error("The popup was closed"));
```
Comment (schettn) [inline src/PopupWindow.js]:
```suggestion
            this.close();
            reject(new Error("The popup was closed"));
```

File: src/PopupWindow.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+           */
+        }
+      }, 500);
+    });
+  }
+
+  cancel() {
+    if (this.iid) {
+      window.clearInterval(this.iid);
+      this.iid = null;
```
Comment (schettn) [inline src/PopupWindow.js]:
```suggestion
      window.clearInterval(this.iid);
      
      this.iid = null;
```

File: src/utils.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
+//#region > Functions
+/**
+ * @returns {values} Params selected from a query
+ * @description This function gets params out of a query
+ */
+function toParams(query) {
```
Comment (schettn) [inline src/utils.js]:
```suggestion
/**
 * Extracts all parameters of a given query string.
 *
 * @function
 * @returns {values} Params selected from a query
 */
function toParams(query) {
```

File: src/utils.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+    return values;
+  }, {});
+}
+
+/**
+ * @returns {query} A query for requests
+ * @description This function builds a query out of params
+ */
+function toQuery(params, delimiter = "&") {
```
Comment (schettn) [inline src/utils.js]:
```suggestion
/**
 * Generates a query string out of given parameters.
 *
 * @function
 * @returns {query} A query for requests
 */
function toQuery(params, delimiter = "&") {
```

File: src/utils.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    return query;
+  }, "");
+}
+
+/**
+ * @returns {string} A guid string
+ * @description This function generates a random guid string
+ */
```
Comment (schettn) [inline src/utils.js]:
```suggestion
/**
 * Generates a random GUID.
 *
 * @function
 * @returns {string} A random GUID
 */
```

File: src/utils.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+function GuidGenerator() {
```
Comment (schettn) [inline src/utils.js]: This function should also be lowercase as all others.
Do not forget to change this in the whole source code.

**Review by schettn — CHANGES_REQUESTED** [review]
Pls check.

**Review by schettn — APPROVED** [review]
LGTM

File: .vscode/settings.json [hunk 16]
```diff
@@ -1,4 +1,15 @@
...
+		"plugins": [
+			"github"
+		],
+		"rules": {
+			"closeAtx": true
+		},
+		"github": {
+			"repository": "https://github.com/mrmlnc/vscode-remark"
+		}
+	},
```
Comment (Aichnerc) [inline .vscode/settings.json:12]: Is this the right indentation?

**Review by Aichnerc — APPROVED** [review]
Looks okay, found something though.

File: .vscode/settings.json (hunk 16 above)
Comment (pinterid) [reply .vscode/settings.json:12]: The indentation only looks like that on GitHub.

File: src/GitHub/index.jsx [hunk 17]
```diff
@@ -0,0 +1,159 @@
+//#region > Imports
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+//> MDB
+// "Material Design for Bootstrap" is a great UI design framework
+import { MDBBtn, MDBIcon } from "mdbreact";
```
Comment (kleberbaum) [inline src/GitHub/index.jsx]:
```suggestion
import React, { Component } from "react";
// Runtime type checking for React props and similar objects.
import PropTypes from "prop-types";
//> MDB
// "Material Design for Bootstrap" is a great UI design framework
import { MDBBtn, MDBIcon } from "mdbreact";
```

File: src/PopupWindow.js [hunk 18]
```diff
@@ -0,0 +1,88 @@
+//#region > Imports
+import { toParams } from "./utils";
```
Comment (kleberbaum) [inline src/PopupWindow.js:3]: Please add a comment.

File: src/GitHub/index.jsx [hunk 19]
```diff
@@ -0,0 +1,159 @@
+//#region > Imports
+import React, { Component } from "react";
```
Comment (kleberbaum) [inline src/GitHub/index.jsx:4]:
```suggestion
//> React
// Contains all the functionality necessary to define React components
import React, { Component } from "react";
```

File: src/GitHub/index.jsx [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+import PropTypes from "prop-types";
```
Comment (kleberbaum) [inline src/GitHub/index.jsx:6]:
```suggestion
// Runtime type checking for React props and similar objects.
import PropTypes from "prop-types";
```

**Review by kleberbaum — CHANGES_REQUESTED** [review]
Should be ok to merge after a few changes.

File: src/GitHub/index.jsx [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+  onRequest: PropTypes.func,
+  onSuccess: PropTypes.func.isRequired,
+  onFailure: PropTypes.func.isRequired,
+  redirectUri: PropTypes.string.isRequired,
+  scope: PropTypes.string,
+};
+//#endregion
+
+//#region > Exports
+//> Default Class
```
Comment (kleberbaum) [inline src/GitHub/index.jsx]:
```suggestion
//> Default Component
```

File: src/utils.js [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+    let r = (d + Math.random() * 16) % 16 | 0;
+
+    d = Math.floor(d / 16);
+
+    return (c === "x" ? r : (r & 0x3) | 0x8).toString(16);
+  });
+}
+//#endregion
+
+//#region > Exports
```
Comment (kleberbaum) [inline src/utils.js:59]:
```suggestion
//#region > Exports
//> Default Functions
```

File: src/index.js [hunk 23]
```diff
@@ -0,0 +1,12 @@
+//#region > Imports
+import GitHubOAuth from "./GitHub";
+//#endregion
+
+//#region > Exports
```
Comment (kleberbaum) [inline src/index.js:5]:
```suggestion
//#region > Exports
//> Default Component
```

File: src/PopupWindow.js [hunk 24]
```diff
@@ -0,0 +1,89 @@
+//#region > Imports
+//> Utils
+import { toParams } from "./utils";
+//#endregion
+
+//#region > Components
```
Comment (kleberbaum) [inline src/PopupWindow.js]:
```suggestion
//#region > Classes
```

**Review by kleberbaum — CHANGES_REQUESTED** [review]
Should be ok to merge after a few changes.

**Review by kleberbaum — APPROVED** [review]
Should be ok to merge.

**Pull request merged**